Modules/_ctypes/callbacks.c

```diff
@@ -146,47 +146,55 @@ static void _CallPythonObject(void *mem,
                               int flags,
                               void **pArgs)
 {
-    Py_ssize_t i;
-    PyObject *result;
-    PyObject *arglist = NULL;
-    Py_ssize_t nArgs;
+    PyObject *result = NULL;
+    PyObject **args = NULL;
+    Py_ssize_t i = 0, j = 0, n_args = 0;
     PyObject *error_object = NULL;
     int *space;
     PyGILState_STATE state = PyGILState_Ensure();
 
-    nArgs = PySequence_Length(converters);
+    n_args = PySequence_Length(converters);
     /* Hm. What to return in case of error?
        For COM, 0xFFFFFFFF seems better than 0.
     */
-    if (nArgs < 0) {
+    if (n_args < 0) {
         PrintError("BUG: PySequence_Length");
         goto Done;
     }
 
-    arglist = PyTuple_New(nArgs);
-    if (!arglist) {
-        PrintError("PyTuple_New()");
-        goto Done;
+    PyObject *args_stack[_PY_FASTCALL_SMALL_STACK];
+    if (n_args <= (Py_ssize_t)Py_ARRAY_LENGTH(args_stack)) {
```

Comment:
Many ctypes function use alloca(). ctypes limits the maximum number of arguments:

```
/*
 * bpo-13097: Max number of arguments _ctypes_callproc will accept.
 *
 * This limit is enforced for the `alloca()` call in `_ctypes_callproc`,
 * to avoid allocating a massive buffer on the stack.
 */
#define CTYPES_MAX_ARGCOUNT 1024
```

On x86-64, it would allocate 8 KiB in the worst case. It sounds acceptable, no?

It would avoid to "waste" args_stack when you need less than _PY_FASTCALL_SMALL_STACK items (5). And it would avoid the "expensive" heap allocation.

Reply:
I added a second check, just to be extra safe, but the existing check is enough: https://github.com/python/cpython/pull/31188